src/Microsoft.Extensions.SecretManager.Tools/Internal/MsBuildProjectReader.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Xml;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Evaluation;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.ProjectModel.Internal;
+using GracefulException = Microsoft.Extensions.SecretManager.Tools.Internal.GracefulException;
+
+namespace Microsoft.Extensions.SecretManager.Tools.Internal
+{
+    internal class MsBuildProjectReader
+    {
+        public string ReadUserSecretId(IFileInfo fileInfo)
+        {
+            if (fileInfo == null || !fileInfo.Exists)
+            {
+                throw new GracefulException($"Could not find file '{fileInfo.PhysicalPath}'");
+            }
+
+            var projectDir = Path.GetDirectoryName(fileInfo.PhysicalPath);
+
+            var dotNetHome = Path.GetDirectoryName(new Muxer().MuxerPath);
+            var sdkResolver = new DotNetCoreSdkResolver(dotNetHome);
+            var sdk = sdkResolver.ResolveProjectSdk(projectDir);
+
+            // workaround limitations in using MSBuild API
+            Environment.SetEnvironmentVariable("MSBUILD_EXE_PATH", Path.Combine(sdk.BasePath, "MSBuild.exe"));
+
+            var projectCollection = new ProjectCollection();
+            Project project;
+            using (var stream = fileInfo.CreateReadStream())
+            {
+                var xmlReader = XmlReader.Create(stream);
+                var xml = ProjectRootElement.Create(xmlReader);
+                xml.FullPath = fileInfo.PhysicalPath;
+
+                var globalProperties = new Dictionary<string, string>
+                {
+                    { "DesignTimeBuild", "true" },
+                    { "MSBuildExtensionsPath", sdk.BasePath }
+                };
+
+                project = new Project(xml, globalProperties, /* toolsVersion: */ null, projectCollection);
+            }
+
+            var userSecretsId = project.Properties.FirstOrDefault(p => p.Name == "UserSecretsId")?.EvaluatedValue;
```

Comment:
I assumed it would throw if the property didn't exist, but the API doc says otherwise, so I'll update to use that API.